File: src/api/monthReview.ts
import { api } from './client'
import type { MonthUpsert, MonthUpsertResult, SpendingMonthUpsert, SpendingUpsertResult } from '../types/api'
import type { MetricEvidence } from '../types/metrics'

export type ReviewState = 'not_started' | 'in_progress' | 'ready_to_review' | 'closed' | 'needs_review' | 'unreviewed_history'
export type ReviewedFeeds = { balances: boolean; spending: boolean; take_home: boolean }
export interface MonthReview {
  month: string
  state: ReviewState
  input_revision: string
  reviewed: ReviewedFeeds
  coverage: ReviewedFeeds & { spending_nonzero: boolean; missing_account_ids: number[]; missing_category_ids: number[] }
  can_close: boolean
  blockers: string[]
  eligible_spending: boolean
  eligible_savings: boolean
  legacy_eligible: boolean
  closed_at: string | null
  closed_by: string | null
  source_link: string
}
export interface MonthReviewList {
  adopted_on: string | null
  default_month: string | null
  months: MonthReview[]
}
export interface MonthSave {
  expected_revision: string
  request_id?: string
  balances?: MonthUpsert
  spending?: SpendingMonthUpsert
  reviewed: ReviewedFeeds
  close: boolean
}
export interface MonthSaveResult {
  month: string
  review: MonthReview
  balances: MonthUpsertResult | null
  spending: SpendingUpsertResult | null
  batch_id: string | null
}
export interface SpendingEvidence {
  month: string | null
  review: MonthReview | null
  metrics: MetricEvidence[]
  comparison: MetricEvidence
  rolling: MetricEvidence
}
export const REVIEW_LABELS: Record<ReviewState, string> = {
  not_started: 'Not started', in_progress: 'In progress', ready_to_review: 'Ready to review',
  closed: 'Closed', needs_review: 'Needs review', unreviewed_history: 'Unreviewed history',
}
export const fetchMonthReviews = () => api<MonthReviewList>('/month-review')
export const fetchMonthReview = (month: string) => api<MonthReview>(`/month-review/months/${month}`)
export const saveMonthReview = (month: string, body: MonthSave) => api<MonthSaveResult>(`/month-review/months/${month}`, { method: 'PUT', body: JSON.stringify(body) })
export const batchCloseMonths = (months: { month: string; expected_revision: string }[]) => api<{ months: MonthReview[]; batch_id: string | null }>('/month-review/batch-close', {
  method: 'POST', body: JSON.stringify({ months, reviewed: { balances: true, spending: true, take_home: true } }),
})
export const fetchSpendingEvidence = (month?: string) => api<SpendingEvidence>(`/metrics/spending${month ? `?month=${month}` : ''}`)

File: backend/app/schemas/month_review.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.net_worth import MonthUpsert, MonthUpsertResult
from app.schemas.spending import SpendingMonthUpsert, SpendingUpsertResult

ReviewState = Literal[
    "not_started", "in_progress", "ready_to_review", "closed", "needs_review", "unreviewed_history"
]


class ReviewedFeeds(BaseModel):
    balances: bool = False
    spending: bool = False
    take_home: bool = False


class FeedCoverage(BaseModel):
    balances: bool
    spending: bool
    take_home: bool
    spending_nonzero: bool
    missing_account_ids: list[int] = []
    missing_category_ids: list[int] = []


class MonthReviewOut(BaseModel):
    month: date
    state: ReviewState
    input_revision: str
    reviewed: ReviewedFeeds
    coverage: FeedCoverage
    can_close: bool
    blockers: list[str]
    eligible_spending: bool
    eligible_savings: bool
    legacy_eligible: bool
    closed_at: datetime | None = None
    closed_by: str | None = None
    source_link: str


class MonthReviewListOut(BaseModel):
    adopted_on: date | None
    default_month: date | None
    months: list[MonthReviewOut]


class MonthSaveIn(BaseModel):
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    request_id: UUID | None = None
    balances: MonthUpsert | None = None
    spending: SpendingMonthUpsert | None = None
    reviewed: ReviewedFeeds = Field(default_factory=ReviewedFeeds)
    close: bool = False


class MonthSaveOut(BaseModel):
    month: date
    review: MonthReviewOut
    balances: MonthUpsertResult | None = None
    spending: SpendingUpsertResult | None = None
    batch_id: UUID | None = None


class BatchCloseMonth(BaseModel):
    month: date
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")


class BatchCloseIn(BaseModel):
    months: list[BatchCloseMonth] = Field(min_length=1, max_length=240)
    reviewed: ReviewedFeeds


class BatchCloseOut(BaseModel):
    months: list[MonthReviewOut]
    batch_id: UUID | None
